src/TransitionSystems/quotient.rs

```diff
@@ -1,37 +1,497 @@
+                for s_transition in &s {
+                    // ϕ_T ∧ Inv(l2_t)[r |-> 0] ∧ Inv(l1_t)
+                    let mut guard_zone = get_allowed_fed(&loc_t, t_transition);
+
+                    // ϕ_T ∧ Inv(l2_t)[r |-> 0] ∧ Inv(l1_t) ∧ ϕ_S ∧ Inv(l2_s)[r |-> 0] ∧ Inv(l1_s)
+                    guard_zone.intersect(&get_allowed_fed(&loc_s, s_transition));
+
+                    let target_locations = merge(
+                        &t_transition.target_locations,
+                        &s_transition.target_locations,
+                    );
+
+                    //Union of left and right updates
+                    let mut updates = t_transition.updates.clone();
+                    updates.append(&mut s_transition.updates.clone());
+
+                    //println!("Rule 1: {}", guard_zone);
+
+                    transitions.push(Transition {
+                        guard_zone,
+                        target_locations,
+                        updates,
+                    });
+                }
+            }
+        }
+
+        //Rule 2
+        if self.S.actions_contain(action) && !self.T.actions_contain(action) {
+            //Independent S
+            for s_transition in &s {
+                let guard_zone = get_allowed_fed(&loc_s, s_transition);
+
+                //println!("Rule 2: {}", guard_zone);
+
+                let target_locations = merge(&loc_t, &s_transition.target_locations);
+                let updates = s_transition.updates.clone();
+                transitions.push(Transition {
+                    guard_zone,
+                    target_locations,
+                    updates,
+                });
+            }
+        }
+
+        if self.S.get_output_actions().contains(action) {
+            // TODO: check with Martijn
+            // new Rule 3 (includes rule 4 by de-morgan)
+            let mut g_s = Federation::empty(self.dim);
+
+            for s_transition in &s {
+                let allowed_fed = get_allowed_fed(&loc_s, s_transition);
+                g_s.add_fed(&allowed_fed);
+            }
+
+            //println!("Rule 3: {}", g_s.inverse());
+
+            // Rule 5
+            let mut inv_l_s = Federation::full(self.dim);
+            loc_s.apply_invariants(&mut inv_l_s);
+
+            //println!("Rule 5: {}", inv_l_s.inverse());
+
+            transitions.push(Transition {
+                guard_zone: (!inv_l_s) + (!g_s),
+                target_locations: universal_location.clone(),
+                updates: reset_all.clone(),
+            });
+        } else {
+            // Rule 5
+            let mut inv_l_s = Federation::full(self.dim);
+            loc_s.apply_invariants(&mut inv_l_s);
+
+            //println!("Rule 5: {}", inv_l_s.inverse());
+
+            transitions.push(Transition {
+                guard_zone: !inv_l_s,
+                target_locations: universal_location.clone(),
+                updates: reset_all.clone(),
+            });
+        }
```

Comment:
Looks a bit weird at first glance with two rule 5s, i would write Rule 5 part 1 and 2 maybe even mention how it relates to rule 8

Reply:
It is because rule 5 either is applied when rule 3 applies or when rule 3 does not apply. They are combined into a single transition as they have the same target location and resets. Cleared it up with a minor comment